runtime-sdk/src/crypto/signature/mod.rs

```diff
@@ -66,6 +159,78 @@ impl PublicKey {
         }
     }
 
+    /// Verify the signature of a message.
+    pub fn verify_by_type(
+        &self,
+        signature_type: SignatureType,
+        context: Option<&[u8]>,
+        message: &[u8],
+        signature: &Signature,
+    ) -> Result<(), Error> {
+        match self {
+            Self::Ed25519(pk) => match signature_type {
+                SignatureType::Ed25519_Oasis => {
+                    let context = context.ok_or(Error::InvalidArgument)?;
+                    pk.verify(context, message, signature)
+                }
+                SignatureType::Ed25519_Pure => pk.verify_raw(message, signature),
+                SignatureType::Ed25519_Sha512 => {
+                    let mut digest = Sha512::new();
+                    digest.update(message);
+                    pk.verify_digest(digest, signature)
+                }
+                _ => Err(Error::InvalidArgument),
+            },
+            Self::Secp256k1(pk) => match signature_type {
+                SignatureType::Secp256k1_Oasis => {
+                    let context = context.ok_or(Error::InvalidArgument)?;
+                    pk.verify(context, message, signature)
+                }
+                SignatureType::Secp256k1_Pure => pk.verify_raw(message, signature),
+                SignatureType::Secp256k1_Keccak256 => {
+                    let mut digest = sha3_0_9_1::Keccak256::new();
+                    digest.update(message);
+                    pk.verify_digest(digest, signature)
+                }
+                _ => Err(Error::InvalidArgument),
+            },
+            _ => Err(Error::InvalidArgument),
+        }
+    }
+
+    /// Verify the signature of a pre-hashed message.
+    pub fn verify_prehashed_by_type(
+        &self,
+        signature_type: SignatureType,
+        hash: &[u8],
+        signature: &Signature,
+    ) -> Result<(), Error> {
+        match self {
+            Self::Ed25519(pk) => match signature_type {
+                SignatureType::Ed25519_Sha512 => {
+                    if hash.len() != <Sha512 as sha2::digest::FixedOutput>::OutputSize::USIZE {
+                        return Err(Error::InvalidArgument);
+                    }
+                    let digest = digests::DummyDigest::<Sha512>::new_precomputed(hash);
+                    pk.verify_digest(digest, signature)
+                }
+                _ => Err(Error::InvalidArgument),
+            },
+            Self::Secp256k1(pk) => match signature_type {
+                SignatureType::Secp256k1_Keccak256 => {
+                    if hash.len() != <sha3_0_9_1::Keccak256 as sha3_0_9_1::digest::FixedOutput>::OutputSize::USIZE {
+                        return Err(Error::InvalidArgument);
+                    }
+                    // Use SHA-256 for RFC6979 even if Keccak256 was used for the message.
+                    let digest = digests::DummyDigest::<sha2::Sha256>::new_precomputed(hash);
```

Comment:
So in this case we don't actually care about what hash function was used to compute the hash as it seems we always use SHA-256 for inner hashing (and Keccak256 is only used for the length check)? Does this mean we could just use a generic `Secp256k1_PreHash` as long as the length matches (or what exactly is the requirement on the input)?